### PR TITLE
 Add cases for Optional Expressions as DestructuringAssignmentTarget

### PR DESCRIPTION
**Pull request by leobalter**
cc @ljharb @bcoe

### CODE REVIEW

File: src/dstr-assignment/array-elem-nested-memberexpr-optchain-prop-ref-init.case [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  type: SyntaxError
+---*/
+
+//- setup
+$DONOTEVALUATE();
+var x = {};
+//- elems
+[x?.y = 42]
+//- vals
+[23]
```
Comment (bcoe) [inline src/dstr-assignment/array-elem-nested-memberexpr-optchain-prop-ref-init.case]: what purpose does `[23]` serve here?

File: src/dstr-assignment/array-elem-nested-memberexpr-optchain-prop-ref.case [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+negative:
+  phase: parse
+  type: SyntaxError
+---*/
+
+//- setup
+$DONOTEVALUATE();
+var x = {};
+//- elems
+[x?.y]
```
Comment (bcoe) [inline src/dstr-assignment/array-elem-nested-memberexpr-optchain-prop-ref.case]: :point_up: does `[x.y]` without a RHS evaluate as a destructuring assignment?

I think I would have assumed that this was an `[ Expression ]`. Would this production work:

```
const arr = [x?.y, x?.z];
```

File: src/dstr-assignment/array-elem-put-obj-literal-optchain-prop-ref.case [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+---*/
+
+//- setup
+$DONOTEVALUATE();
+//- elems
+[{
+  set y(val) {
+    throw new Test262Error('The property should not be accessed.');
+  }
+}?.y]
```
Comment (bcoe) [inline src/dstr-assignment/array-elem-put-obj-literal-optchain-prop-ref.case]: same question as above, probably me failing to read this production properly.

**Comment by rkirsling** [comment]
I'm not sure if some of these tests were correct?

None of these are a SyntaxError with `.`, so I don't believe `?.` should change that:
https://github.com/tc39/test262/blob/master/test/language/statements/for-of/dstr/obj-prop-elem-target-obj-literal-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-of/dstr/array-elem-nested-memberexpr-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-of/dstr/obj-prop-elem-target-memberexpr-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-of/dstr/array-elem-put-obj-literal-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-in/dstr/obj-prop-elem-target-obj-literal-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-in/dstr/obj-prop-elem-target-memberexpr-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-in/dstr/array-elem-put-obj-literal-optchain-prop-ref.js
https://github.com/tc39/test262/blob/master/test/language/statements/for-in/dstr/array-elem-nested-memberexpr-optchain-prop-ref.js

**Comment by leobalter** [comment]
@rkirsling I edited your comment to fix the links

**Pull request merged**